design/one-pager-configuration-stacks.md

```diff
@@ -288,6 +283,10 @@ for the resources:
     environment by just creating a CR instance instead of having everything in
     local.
   * Stacks do a better job for declaring CRD dependencies and UI annotation metadata.
+* Using kustomize or Go templates could save us from adding a Go function for each
+  kind we add even though this'd limit the future addition abilities it'd could
+  be preferrable.
```

Comment:
I have an inkling that using kustomize could be elegant here. Something like:

* User provides the base YAML for a bunch of resources.
* Our controller generates an overlay that sets resource names,  labels, and cross-resource reference.
* Our controller applies the generated overlay to the base layer and applies it to the API server.

This could allow folks to fork and define their own minimal/easy stacks simply by writing YAML (and stack scaffolding metadata). It seems like there's a bunch of potential overlap with template stacks here, but I don't think that needs to block this effort. It could be fine to write a purpose-built controller that uses Kustomize to do this specific task, then assess replacing it with template stacks if and when they're ready.

Reply:
My initial intent was to just use Kustomize for all transformations and let the controller be generic. However, I had gotten stuck while trying to do it, especially the part where you declare a reference to be filled by X kind that exists in the pack with that name.

I went through a few examples, dug deeper and finally found that there is a support for CRDs to declare custom reference fields. https://github.com/kubernetes-sigs/kustomize/blob/master/examples/transformerconfigs/crd/README.md

I have converted the minimal GCP YAML files into that structure and it works very well. We move the burden of the reference handling from the controller to the `kustomizeconfig.yaml` file. Note that we still need to add some lines to that file for each _kind_ that we add. But I guess it's better than adding a Go function.

In fact, we can have a giant `kustomizeconfig.yaml` file in `stack-gcp` that declares the dependencies for each type and fetch it, so, author will not worry about the reference bindings at all.

The controller will use the `kustomization.yaml` file that has `apiVersion: kustomize.config.k8s.io/v1beta1 kind: Kustomization` metadata. This introduces a dependency on kustomize but as far as I see, it's already pretty much the standard in the ecosystem for configuration management and templating. It's even in `kubectl`. So, it should be fine to depend on it IMO.

@prasek what do you think? The controller will be generic but the developer experience will change from dumping all yamls to dumping a kustomize directory and we'll depend on kustomize. Keep in mind that this is pretty much an implementation detail, nothing will change for clients that create the CR, i.e. Upbound backend.

Reply:
@negz There is one caveat: the custom transformer config identifies resources with only `name` and `kind`, it doesn't include `apiVersion` during its search. So, the resources that share the same `kind` but has different `apiVersion`, i.e. multicloud conf stacks, will be treated in the same way.

This would cause problem only if say the stack has one AWS Provider and GCP Provider that share the same name. For other resources, most of the time `kind` is different, like VPC is named `VPC` in AWS but `Network` in GCP. So, it should be fine except for `Provider` kinds which share the same kind across infra stacks.

Opened https://github.com/kubernetes-sigs/kustomize/issues/1885 to track this.

Reply:
@muvaf using a kustomize directory sounds good and is aligned with the template stacks direction towards pluggable templating/config engines including kustomize support, so this direction makes sense.